### PR TITLE
feat: add public api headers

### PR DESCRIPTION
**Pull request by anonrig**

### CODE REVIEW

File: include/ada/url.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    /**
+     * A URL’s port is either null or a 16-bit unsigned integer that identifies a networking port. It is initially null.
+     */
+    std::optional<uint16_t> port;
+
+    /**
+     * A URL’s path is either an ASCII string or a list of zero or more ASCII strings, usually identifying a location.
+     * TODO: Change this to list.
+     */
+    char* path;
```
Comment (anonrig) [inline include/ada/url.h:40]: Specification defines path as either a string or a list of zero or more ASCII strings.

**Comment by lemire** [comment]
A convenient way to represent strings without allocation is `string_view`... It can be better than a C string in some instances. In either case, you need to have clear rules about who allocates and deallocate the memory.

File: src/parser.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+#define ADA_PARSER_H
+
+#include <optional>
+
+namespace ada {
+
+  class Parser {
+
+  public:
+    Parser(const char* input, std::optional<ada::URL> optional_base_url, std::optional<ada::encoding_type> encoding, std::optional<ada::state> state);
```
Comment (lemire) [inline src/parser.h:11]: You take in a `const char*` input and then cast away the `const` qualifier.  Some people will object to such an approach.

Comment (anonrig) [reply src/parser.h]: I converted this to `string_view`. I believe even though `const std::string_view`, we do not need to drop the const  qualifier, and now can just iterate through the `const std::iterator`.

**Comment by anonrig** [comment]
> A convenient way to represent strings without allocation is `string_view`... It can be better than a C string in some instances. In either case, you need to have clear rules about who allocates and deallocate the memory.

Does this also apply to `URL` struct?

**Comment by lemire** [comment]
All I am saying is that one should consider using string_view in some instances. It uses more memory than a mere pointer (since it also tracks the length), but having access to the length in constant time is also a benefit. In practice, it tends to generate very fast code.

**Comment by lemire** [comment]
(This is not to say that C strings are obsolete!!!)

**Review by lemire — APPROVED** [review]

**Pull request merged**